### ISSUE TITLE
Extensions sandbox policy

### ISSUE THREAD

**Issue by cdowdy** [bolt/bolt#7740]
Details
-------  

extensions after 3.3+ have to whitelist tags used by the extensions template/twig. 

| Question                      | Answer
|----------------- | ------
| Relevant Bolt Version | legacy / 3.5 / 3.6 / master
| Install type                   |  Composer install 
| BC Break                     | yes
| PHP version                 | 7.2
| Web server                  | Nginx 1.15
  

Reproduction
------------

install an extension that can use "user generated content" - ie allows an image to be used in a blog post creation - like https://market.bolt.cm/view/cdowdy/betterthumbs
  
in a default install of bolt with default generated records add the following to the ``pages`` contenttype's body HTML field type: ``allowtwig: true``  https://docs.bolt.cm/3.6/fields/html#options

```yaml  
        body:
            type: html
            allowtwig: true  
```

create or edit a "page" contenttype that includes an image from bolt's "files" in the "body content" and the extensions "twig tag" like the following for betterthumbs:  

```html  
{{ img( '/alarm-clock-gold-hands-of-a-clock-1778.jpg' ) }} 
```  

![2019-02-12_11-37-08](https://user-images.githubusercontent.com/5969974/52651651-8b74e780-2eba-11e9-97b9-e7612a18c665.png)  

save the contenttype and then view saved version on the site. 

### Bug summary 

with user entered content you're greeted with ``Uncaught Exception: Twig_Sandbox_SecurityNotAllowedTagError: Twig_Sandbox_SecurityNotAllowedTagError in SecurityPolicy.php line 158: `` 

I looked at the changelog for bolt 3.3 and couldn't find any mentions of the sandbox being changed: 

https://github.com/bolt/bolt/blob/3.3/changelog.md  
https://bolt.cm/item/bolt-3-3-0-release-candidate-is-out-now

but with mentions of it at the following but they don't deal with extensions or with whitelisting tags:  
https://github.com/bolt/bolt/pull/6127  
https://github.com/bolt/bolt/pull/6131  
https://github.com/bolt/bolt/pull/6652/files  
https://github.com/bolt/bolt/pull/6812  

and the seo extension https://github.com/bolt/boltforms/issues/199  

I also looked at the documentation and there are no mentions of whitelisting official twig tags in the extension https://docs.bolt.cm/3.6/extensions/introduction , bolts extension starter, starter extended or stencil. 

### Expected result
extensions to keep working and display images entered into a post Or docs stating "you must whitelist your tags, and all tags that you'll use that come standard with twig unless I've skipped over something :) which is always very, very possible haha.  

### Actual result

twig sandbox error.  

So do we have to author our extensions now and keep a whitelist of tags that we'll use or the end user might use? or just the tags we'll use?

**Comment by cdowdy** [comment]
hey all just checking is to see if anyone has had a chance to look at this or tell me I've missed something obvious haha.. Thanks!

**Comment by cdowdy** [comment]
Hey again just seeing if anyone has had a chance to look at this.

**Comment by cdowdy** [comment]
> Hey again just seeing if anyone has had a chance to look at this

so I'm guessing thats a no

**Comment by SvanteRichter** [comment]
This was a breaking change and should not have been done in a minor release.

To be clear: the new way *is the right way*, but it should not have been changed without a major bump and with the corresponding release notes.

There was a lot of discussion about this way back when, it was also quite turbulent so unfortunately nothing came out of it.

**Comment by cdowdy** [comment]
hey thanks @SahAssar ! ..  

Whats "the new way"? 

I've gone over the extension [starter](https://github.com/bolt/bolt-extension-starter/), [starter extended](https://github.com/bolt/bolt-extension-starter-extended/) , and [stencil](https://github.com/bolt/bolt-extension-stencil/), the bolt.cm extension [docs for registering twig functions/filters](https://docs.bolt.cm/3.6/extensions/intermediate/twig), docs for [twig runtimes](https://docs.bolt.cm/3.6/extensions/intermediate/twig-runtimes) and dont see anything that has changed from the previous docs or anything that is different between those docs/examples and other extensions that allow you to use them in content. 

I can't find any documentation, notes, commits or anything related to changing how our extensions function or new ways to do it. I could very well be skipping over it though!

**Comment by SvanteRichter** [comment]
The "new way" is to force every extension author to consider if their twig functions are safe or not. To be clear, that has not been done in a good way at all (should have bumped major versions, should have updated docs, should have had a migration path, should have updated examples and so on), but the old way was making too many unsafe functions available within the sandbox and did not work as well with the twig system (I'm regurgitating from memory here, this was like 1-2 years ago IIRC).

So basically I'm saying this shouldn't have been done in a 3.X, but perhaps in a 4.X.

**Comment by cdowdy** [comment]
Alright then. Yeah it was from last June. Wasn't aware there was a new way since it's not mentioned anywhere or what exactly to do or how to remedy it. 

This at least gives me something to tell people and point to when I'm asked why my extensions have broken their site and made it unusable and that there's no fix.

**Comment by SvanteRichter** [comment]
We had multiple heated discussions around backwards compat around that time. I was usually on the side of maintaining backwards compat, but I can't remember how this one went. Anyway, the main reason I responded was that this popped up in my github notifications right now. If you want an up to date answer it'd be better to talk to @bobdenotter .

**Comment by cdowdy** [comment]
Nah. Thanks @SahAssar ! Much appreciated finally getting something.
